Add option to disable dual reductions for Clp

patrickvossler18 opened the pull request:
This pull request allows users of the C interface to prevent Clp from using dual reductions during the presolving stage by specifying a reduction type when calling `Cbc_solve`. While it should be possible for users to disable presolving completely using `Cbc_setParameter`, it appears there is [an issue](https://github.com/coin-or/Cbc/issues/512) with such settings being passed to Clp from Cbc.  

Furthermore, this option is critical for users solving MIPs using lazy constraints. If one cannot disable dual reductions or disable presolving altogether then the presolving stage will delete variables such that the constraints that are inserted by the constraint handler lead to infeasibility. 

Finally, to preserve backward compatibility, the reduction type is set to a default value that does not change the Clp options by default. If the maintainers have a preferred way for ensuring that the default behavior is the same, please let me know and I can modify my PR.

patrickvossler18 committed:
Add option to disable dual reductions for Clp

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/coin-or/Cbc?pullRequest=630) <br/>All committers have signed the CLA.

patrickvossler18 committed:
add missing argument in Cbc_solve initialization

patrickvossler18 committed:
change how we pass the reduction flag

- add a method for disabling dual reductions to Cbc_Model
- Remove reduction type argument from Cbc_solve and Cbc_solveLinearProgram
- Change logic in Cbc_solveLinearProgram to depend on the value of the red_type parameter in Cbc_Model

patrickvossler18 commented:
Looks like the checks are failing for Windows i686 because of an issue in the setup of the github action. This seems to be happening in other COIN-OR repos too https://github.com/coin-or/Clp/pull/284

jhmgoossens commented:
Thanks for this PR!

Yes, the failed Windows i686 check has nothing to do with your PR (See [Issue 25](https://github.com/coin-or/COIN-OR-OptimizationSuite/issues/25) on OptimizationSuite).

Looking at name of the added method "Cbc_**disable**DualReds" and the default option to "LPR_Default = 0, /*! Solver will **use** default presolve transformations */", would a more descriptive name for the method not be "Cbc_setDualReductionsType", or similar?
Also, an added test in [test/CInterfaceTest.c](https://github.com/coin-or/Cbc/blob/master/test/CInterfaceTest.c) would be nice. 

I don't use the C Interface myself. Other than the name of the added method and a test, I don't have comments.
This PR offers a workaround to an existing issue. Looking at recent commits, could maybe @jjhforrest, @samuelsbrito or @tkralphs look at this PR?

patrickvossler18 committed:
Change function name to Cbc_setDualReductionsType

- Add getter function

patrickvossler18 committed:
Add simple test for LP dual reduction type

patrickvossler18 commented:
Thanks for looking through my PR. I like your suggestion for changing the function name so I've made that change, added a function to get the dual reduction value, and added a simple test to confirm that we are successfully setting reduction value.
